File: 6.Maximum_Subarray/sol.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Maximum Subarray (Go Version)
// ======================================================================
// Description:
// Given an integer array `nums`, find the subarray with the largest sum,
// and return its sum.
//
// 📋 Rules:
// 1. A subarray is a contiguous part of an array.
// 2. Return the maximum sum found.
// 3. Handle cases with negative numbers correctly.
//
// 💡 Examples:
// - MaxSubArray([-2,1,-3,4,-1,2,1,-5,4]) => 6
// - MaxSubArray([1])                     => 1
// - MaxSubArray([5,4,-1,7,8])            => 23
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)

/**
 * Method: Kadane's Algorithm
 * Complexity: Time O(N) | Space O(1)
 */
func MaxSubArray1(nums []int) int {
	// Initialize with the first element
	currentSum := nums[0]
	maxSum := nums[0]

	for i := 1; i < len(nums); i++ {
		num := nums[i]

		// Logic: curSum = max(num, curSum + num)
		// If adding the previous sum makes the result smaller than num itself
		// (which happens if curSum is negative), we discard the previous sum.
		if currentSum < 0 {
			currentSum = num
		} else {
			currentSum += num
		}

		// Update global max
		if currentSum > maxSum {
			maxSum = currentSum
		}
	}

	return maxSum
}

// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func MaxSubArray(nums []int) int {
	// TODO: Implement your solution here.
	return 0
}

// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

func main() {
	runTests()
}

type TestCase struct {
	nums     []int
	expected int
}

func runTests() {
	testCases := []TestCase{
		{[]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}, 6},
		{[]int{1}, 1},
		{[]int{5, 4, -1, 7, 8}, 23},
		{[]int{-1}, -1},     // Edge case: single negative number
		{[]int{-2, -1}, -1}, // Edge case: all negatives
	}

	fmt.Printf("\n🧪 Testing your [MaxSubArray] function...\n\n")

	header := fmt.Sprintf("%-30s | %-10s | %-10s | Status", "Input nums", "Expected", "Actual")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		result := MaxSubArray(tc.nums)

		isMatch := result == tc.expected
		statusIcon := "✅ PASS"
		if !isMatch {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		// Helper to format slices
		fmtSlice := func(s []int) string {
			return fmt.Sprintf("%v", s)
		}

		numsStr := fmtSlice(tc.nums)
		if len(numsStr) > 28 {
			numsStr = numsStr[:25] + "..."
		}

		fmt.Printf("%-30s | %-10v | %-10v | %s\n",
			numsStr, tc.expected, result, statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func MaxSubArray(nums []int) int {",
		"\t// TODO: Implement your solution here.",
		"\treturn 0",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion

File: 6.Maximum_Subarray/sol.py
import os
import sys
from typing import List

# ======================================================================
# 🧠 CHALLENGE: Maximum Subarray (Python Version)
# ======================================================================
# Description:
# Given an integer array `nums`, find the subarray with the largest sum,
# and return its sum.
#
# 📋 Rules:
# 1. A subarray is a contiguous part of an array.
# 2. Return the maximum sum found.
# 3. Handle cases with negative numbers correctly.
#
# 💡 Examples:
# - max_sub_array([-2,1,-3,4,-1,2,1,-5,4]) => 6  ([4,-1,2,1])
# - max_sub_array([1])                     => 1
# - max_sub_array([5,4,-1,7,8])            => 23
# ======================================================================


# region [📚 Reference Solutions] (Solutions hidden as requested)
def max_sub_array1(nums: List[int]) -> int:
    """
    Method: Kadane's Algorithm
    Complexity: Time O(N) | Space O(1)
    """
    # Initialize with the first element (handles array of size 1)
    current_sum = nums[0]
    max_sum = nums[0]

    # Iterate starting from the second element
    for num in nums[1:]:
        # Decision: Start new subarray at num, or extend previous sum?
        # If current_sum is negative, adding it decreases value, so we restart at num.
        current_sum = max(num, current_sum + num)

        # Update global max
        max_sum = max(max_sum, current_sum)

    return max_sum


# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def max_sub_array(nums):
    # TODO: Implement your solution here.
    return 0
# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def max_sub_array(nums):\n",
        "    # TODO: Implement your solution here.\n",
        "    return 0\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)
        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ([-2, 1, -3, 4, -1, 2, 1, -5, 4], 6),
        ([1], 1),
        ([5, 4, -1, 7, 8], 23),
        ([-1], -1),  # Edge case: single negative number
        ([-2, -1], -1),  # Edge case: all negatives
    ]

    print(f"\n🧪 Testing your [max_sub_array] function...\n")

    header = f"{'Input nums':<30} | {'Expected':<10} | {'Actual':<10} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for nums, expected in test_cases:
        try:
            result = max_sub_array(nums)
        except Exception as e:
            result = "Error"

        is_match = result == expected
        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        # Format output
        nums_str = str(nums)
        if len(nums_str) > 28:
            nums_str = nums_str[:25] + "..."

        print(
            f"{nums_str:<30} | {str(expected):<10} | {str(result):<10} | {status_icon}"
        )

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")

# endregion
